nmxact/xact/image.go

```diff
@@ -185,33 +209,189 @@ func nextImageUploadReq(s sesn.Sesn, upgrade bool, data []byte, off int, imageNu
+		t.MaxRxOff = int64(irsp.Off)
+	}
+	if c.ProgressCb != nil {
+		c.ProgressCb(c, irsp)
+	}
+	if t.TuneWS && t.WCap < IMAGE_UPLOAD_MAX_WS {
+		atomic.AddInt64(&t.WCap, 1)
+		cmp += 1
+	}
+	atomic.AddInt64(&t.WCount, -1)
+
+	// Indicate transition from window being full to with open slot(s)
+	if wFull && (t.WCap-t.WCount == cmp) {
+		return true
+	} else {
+		return false
+	}
+}
+
+func (t *ImageUploadIntTracker) HandleError(off int, err error) bool {
+	/*XXX: there could be an Unauthorize or EOF error  when the rate is too high
+	  due to a large window, we retry. example:
+	  "failed to decrypt message: coap_sec_tunnel: decode GCM fail EOF"
+	  Since the error is sent with fmt.Errorf() API, with no code,
+	  the string may have to be parsed to know the particular error */
+	var wFull = false
+	if t.WCap == t.WCount {
+		wFull = true
+	}
+
+	if t.WCount > IMAGE_UPLOAD_START_WS+1 {
+		atomic.AddInt64(&t.WCap, -1)
+	}
+	t.TuneWS = false
+	atomic.AddInt64(&t.WCount, -1)
+	t.UpdateTracker(off, IMAGE_UPLOAD_STATUS_MISSED)
+
+	// Indicate transition from window being full to with open slot(s)
+	if wFull && t.WCount == t.WCap-1 {
+		return true
+	} else {
+		return false
+	}
+}
+
 func (c *ImageUploadCmd) Run(s sesn.Sesn) (Result, error) {
 	res := newImageUploadResult()
+	ch := make(chan int)
+	done := make(chan int)
+	rspc := make(chan nmp.NmpRsp, IMAGE_UPLOAD_MAX_WS)
+	errc := make(chan error, IMAGE_UPLOAD_MAX_WS)
+
+	t := ImageUploadIntTracker{
+		TuneWS:   true,
+		WCount:   0,
+		WCap:     IMAGE_UPLOAD_START_WS,
+		Off:      c.StartOff,
+		RspMap:   make(map[int]int),
+		MaxRxOff: 0,
+	}
+
+	for int(atomic.LoadInt64(&t.MaxRxOff)) < len(c.Data) {
+		// Block if window is full
+		if !t.CheckWindow() {
+			ch <- 1
+		}
 
-	for off := c.StartOff; off < len(c.Data); {
-		r, err := nextImageUploadReq(s, c.Upgrade, c.Data, off, c.ImageNum)
-		if err != nil {
-			return nil, err
+		t.Mutex.Lock()
+		t.ProcessMissedChunks()
+		t.Mutex.Unlock()
+
+		if t.Off == len(c.Data) {
+			continue
 		}
 
-		rsp, err := txReq(s, r.Msg(), &c.CmdBase)
+		t.Mutex.Lock()
+		r, err := nextImageUploadReq(s, c.Upgrade, c.Data, t.Off, c.ImageNum)
 		if err != nil {
+			t.Mutex.Unlock()
 			return nil, err
 		}
-		irsp := rsp.(*nmp.ImageUploadRsp)
 
-		off = int(irsp.Off)
-
-		if c.ProgressCb != nil {
-			c.ProgressCb(c, irsp)
-		}
+		t.Off = (int(r.Off) + len(r.Data))
 
-		res.Rsps = append(res.Rsps, irsp)
-		if irsp.Rc != 0 {
+		// Use up a chunk in window
+		atomic.AddInt64(&t.WCount, 1)
+		err = txReqAsync(s, r.Msg(), &c.CmdBase, rspc, errc)
+		if err != nil {
+			log.Debugf("err txReqAsync %v", err)
+			t.Mutex.Unlock()
 			break
 		}
+		// Mark the expected offset in successful tx of this chunk. i.e off + len
+		t.UpdateTracker(int(r.Off)+len(r.Data), IMAGE_UPLOAD_STATUS_EXPECTED)
+		t.Mutex.Unlock()
+
+		go func(off int) {
+			select {
+			case err := <-errc:
+				t.Mutex.Lock()
+				sig := t.HandleError(off, err)
+				t.Mutex.Unlock()
```

Comment:
Totally up to you, but you might consider having `HandleError()` do the locking rather than the caller.  Since the mutex always has to be locked when this function executes, that just seems more appealing to me (you can also use `defer` to unlock then, which is a bit safer, all things being equal).

(Same comment applies to `HandleResponse()`).

Reply:
Changed as per comment.

Reply:
Updated as per comment.